### ISSUE TITLE
Array module review followup: Should isRectangularArr, isIreggularArr, etc. be moved to be methods on array type? 

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#18086]
This is a followup item for the [Array module review](https://github.com/Cray/chapel-private/issues/2241).

One item we wanted further discussion on is:

whether or not we should move `isRectangularArr()`, `isIrregularArr()` etc. to be functions on the array type.  Currently these are free functions.

One justification for them being free (rather than on the type itself) is in a generic function you may want to see if you want to query whether some unknown datatypes is a rectangular array and you don't want a compile-time error if the underlying type isn't an array. Of course a simple workaround would be to just do this: `if(isArray(A) && A.isRectangular)`

**Comment by mppf** [comment]
IIRC in the review, somebody pointed out that `isRectangularArr(int)` etc doesn't currently work (i.e. these functions assume the argument is an array). As a result I think it makes sense to make them functions on the array type.

**Comment by stonea** [comment]
Yes, it looks like isRectangularArr(int) gives the following error; so yeah "we should keep isRectangularArr as a free function as a convenience to writing generic functions" seems like a weaker argument.

```
foo.chpl:1: In module 'foo':
foo.chpl:1: error: unresolved call 'isRectangularArr(type int(64))'
$CHPL_HOME/modules/internal/ChapelArray.chpl:933: note: this candidate did not match: isRectangularArr(a: [] ) [120386]
foo.chpl:1: note: because actual argument #1 is a type
$CHPL_HOME/modules/internal/ChapelArray.chpl:933: note: but is passed to non-type formal 'a: [] [120383]'
foo.chpl:1: note: unresolved call had id 211673
./doit: line 8: ./foo_real: No such file or directory
```

**Comment by lydia-duncan** [comment]
I'm in favor of turning these into methods

**Comment by bradcray** [comment]
I also liked the proposal to turn them into methods that Michael and Lydia are referring to above.

**Comment by stonea** [comment]
Ok, seems like this is relatively uncontroversial then. If we do this for the Arr functions I think we would also want to do the same for the Dom functions:

Just to be very specific the current proposal is:

Remove the following free functions and instead have the corresponding methods in the array\domain type:
* `isRectangularDom()` -> `Domain.isRectangular()`
* `isRectangularArr()` -> `Array.isRectangular()`
* `isIrregularDom()` -> `Domain.isIrregular()`
* `isIrregularArr()` -> `Array.isIrregular()`
* `isAssociativeDom()` -> `Domain.isAssociative()`
* `isAssociativeArr()` -> `Array.isAssociative()`
* `isSparseDom()` -> `Domain.isSparse()`
* `isSparseArr()` -> `Array.isSparse()`
